rfcs/20241015-result-accuracy.md

```diff
@@ -0,0 +1,74 @@
+# [RFC] Add result accuracy to transcendental unary ops
+
+Status: In Review<br/>
+Initial version: 10/15/2024<br/>
+Last updated: 10/15/2024<br/>
+Discussion thread:
+
+## Overview
+This RFC proposes adding a new attribute `result_accuracy` to the following transcendental unary ops: `exp`, `expm1`, `log`, `logp1`, `logistic` and `tanh`.
+`result_accuracy` allows the user to choose the implementation of these ops based on the accuracy they request. The choice of implementation is restricted to F32 inputs. 
+
+## Background
+Transcendental ops can have multiple implementations on a processing unit that vary in accuracy and performance.
+By allowing users to select implementation based on accuracy per op, we can offer more tools to weigh tradeoffs between accuracy and performance and also ensure more consistent numerical behaviors across different devices.
+
+### Caveats on targets
+The proposed result_accuracy attribute can only be supported for targets that have multiple implementations of the ops.
+For example, for XLA-CPU, the implementations of these ops are dependent on LLVM and each target may have a different implementation with a different level of accuracy.
+Thus, further analysis needs to be done to support this feature on CPUs. This analysis could be performed while building the compiler, assuming the compiler will only be used on a single type of CPU.
+
+## Proposed Specification
+
+### `result_accuracy`
```

Comment:
@jtristan I'm curious how you feel about speccing things as structs vs the mostly-inlined representation we've been using in StableHLO, as long as the grammar is trivially parsable to you have any preference?

I.e. I could see a spec for `result_accuracy<atol,trol,ulps,mode>` being done as:
1. Adding a new "type" to the spec called `ResultAccuracy` and to each of the proposed ops (`exp`, `expm1`, `log`, `logp1`, `logistic` and `tanh`) adding a single input `(I2) result_accuracy : ResultAccuracy` and the type has a constraint `(C1) if mode != TOLERANCE: atol = rtol = ulps = 0`, or
2. Speccing these as 4 flat arguments `(I2) atol, (I3) rtol, (I4) ulps, (I5) mode` on each of the ops, and adding a constraint to each op as well `(C1) if mode != TOLERANCE: atol = rtol = ulps = 0`

Option (1) feels more correct, but it's less standard in the current spec so curious if you have thoughts on any implications of doing things this way.

Reply:
(cc @sdasgup3, @ghpvnist, @abhigunj)

Reply:
As long as `mode` is optional, I don't see any implications from the IR standpoint. I also prefer option 1 since it looks cleaner.

Reply:
I'm also favorable to option 1.